### PR TITLE
Fb 6174 save run api

### PR DESCRIPTION
**Pull request by labkey-ankurj**

### CODE REVIEW

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java [hunk 1]
```diff
@@ -6534,4 +6536,44 @@ public Object execute(Object o, BindException errors) throws Exception
             return ret;
         }
     }
+
+    @Marshal(Marshaller.Jackson)
+    @RequiresPermission(ReadPermission.class)
+    public class LoadAssayRunAction extends ReadOnlyApiAction<LoadAssayRunForm>
```
Comment (labkey-kevink) [inline experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: As we discussed, let's create GetAssayRunsAction similar to GetAssayBatchesAction instead.

File: study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java [hunk 2]
```diff
@@ -319,6 +323,23 @@ public void testSaveBatchWithAdHocProperties() throws IOException, URISyntaxExce
...
 
+    @Test
+    public void testSaveRunApi() throws IOException, CommandException
+    {
+        Run run = new Run();
+        run.setName("testRun");
+        run.setProperties(Map.of("name","testIntField" ,"rangeURI", "int"));
+        SaveAssayRunCommand saveAssayRunCommand = new SaveAssayRunCommand(SaveAssayBatchCommand.SAMPLE_DERIVATION_PROTOCOL, List.of(run));
+        SaveAssayRunResponse saveAssayRunResponse = saveAssayRunCommand.execute(createDefaultConnection(false), getProjectName());
+        String addedRunLsid = saveAssayRunResponse.getRuns().get(0).getLsid();
```
Comment (labkey-kevink) [inline study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java]: Need to verify the property was included in the response.

File: study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java [hunk 3]
```diff
@@ -319,6 +323,23 @@ public void testSaveBatchWithAdHocProperties() throws IOException, URISyntaxExce
...
+
+        LoadAssayRunCommand loadAssayRunCommand = new LoadAssayRunCommand(addedRunLsid);
+        LoadAssayRunResponse loadAssayRunResponse = loadAssayRunCommand.execute(createDefaultConnection(false), getProjectName());
+        String resultLsid = loadAssayRunResponse.getRun().getLsid();
```
Comment (labkey-kevink) [inline study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java]: Need to verify the property was included in the response.

File: study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java [hunk 4]
```diff
@@ -319,6 +323,23 @@ public void testSaveBatchWithAdHocProperties() throws IOException, URISyntaxExce
         assertEquals("Ad hoc property not found." , propertyURIS.get(0).getPropertyURI(), addedPropertyURIs.get(0));
     }
 
+    @Test
+    public void testSaveRunApi() throws IOException, CommandException
+    {
+        Run run = new Run();
+        run.setName("testRun");
+        run.setProperties(Map.of("name","testIntField" ,"rangeURI", "int"));
```
Comment (labkey-kevink) [inline study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java]: The property map here should have a propertyURI key with a value.  The propertyURI should be one from of the Vocabulary properties.  I'd also validate that you can use a property from a different domain or a property without a domain.  For a property in a domain, you could use the GetDomainAction to get the propertyURI for the 'User' field of the UserAuditDomain'.  For a property outside a domain, a example is "urn:exp.labkey.org/#Comment".

File: experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java (hunk 1 above)
Comment (labkey-ankurj) [reply experiment/src/org/labkey/experiment/controllers/exp/ExperimentController.java]: changed locally.

File: study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java [hunk 5]
```diff
@@ -263,6 +271,54 @@ private Batch getBatch(Connection connection, int batchId) throws IOException, C
...
+        prop2.setRangeURI("int");
+        prop2.setName("testIntField");
+
+        adHocProperties.add(prop1);
+        adHocProperties.add(prop2);
+
+        CreateDomainCommand domainCommand = new CreateDomainCommand("Vocabulary", "TestVocabulary");
+        domainCommand.getDomainDesign().setDescription("Test Ad Hoc Properties");
+        domainCommand.getDomainDesign().setFields(adHocProperties);
+        domainCommand.setColumns(List.of(Map.of("name","testIntField" ,"rangeURI", "int"),
```
[the remark this answers is not in the document]
Comment (labkey-ankurj) [reply study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java]: I should have removed setFields(). It wasn't filling in I guess so I had to use setColumns().

File: study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java (hunk 4 above)
Comment (labkey-ankurj) [reply study/test/src/org/labkey/test/tests/experiment/ExperimentAPITest.java]: How can I get the UserAuditDomain ? Do I need to create one?

File: experiment/src/org/labkey/experiment/controllers/property/PropertyController.java [hunk 6]
```diff
@@ -1159,32 +1176,34 @@ public NavTree appendNavTrail(NavTree root)
...
         @Override
-        public Object execute(ContainerDomainForm containerDomainForm, BindException errors) throws Exception
+        public Object execute(ContainerDomainForm containerDomainForm, BindException errors)
         {
             includeFields = containerDomainForm.isIncludeFields();
             includeProjectAndShared = containerDomainForm.isIncludeProjectAndShared();
 
-            return listDomains(getContainer(), getUser(), containerDomainForm, includeProjectAndShared);
+            //resetting mapper when incoming request is json to capture the above scenario
+            _mapper = null;
```
Comment (labkey-kevink) [inline experiment/src/org/labkey/experiment/controllers/property/PropertyController.java]: As we discussed, let's add a `createResponseObjectMapper` and used it in the `createResponseWriter` method in BaseApiAction.

File: experiment/src/org/labkey/experiment/controllers/property/PropertyController.java [hunk 7]
```diff
@@ -122,13 +124,28 @@
 
     public static final String UNRECOGNIZED_FILE_TYPE_ERROR = "Unrecognized file type. Please upload a .xls, .xlsx, .tsv, .csv or .txt file";
 
+    private static SimpleBeanPropertyFilter _gwtDomainPropertiesFilter;
```
[the remark this answers is not in the document]
Comment (labkey-ankurj) [reply experiment/src/org/labkey/experiment/controllers/property/PropertyController.java]: Removed

File: experiment/src/org/labkey/experiment/controllers/property/PropertyController.java (hunk 6 above)
Comment (labkey-ankurj) [reply experiment/src/org/labkey/experiment/controllers/property/PropertyController.java]: Done

**Pull request merged**